docs/spec_draft.md

```diff
@@ -1268,6 +1269,89 @@ More formally, `result[i0, ..., iR-1]` is equal to:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce
+
+### Semantics
+
+Performs a reduction function `body` over the `inputs` and `init_values` along
+the `dimensions` and produces a `result` tensor.
+
+The order of reductions is implementation-defined, which means that `body` and
+`init_values` must form a monoid to guarantee that the operation produces the
+same results for all inputs on all implementations.
+
+However, this condition doesn't hold for many popular reductions. E.g.
+floating-point addition for `body` and zero for `init_values` don't actually
+form a monoid because floating-point addition is not associative. What this
+means for numeric precision is implementation-defined.
+
+More formally, `results[:][result_index] = reduce(input_slices)` where:
+  * `input_slices` = `inputs[:][ri0, ..., :, ..., riR-1]`, where `ri` are
+    individual elements in `result_index` and `:` are inserted at `dimensions`.
+  * `reduce(input_slices)` = `eval(schedule)` for some binary tree `schedule`
+    where:
+    * `eval(node)` = `body(eval(node.left), eval(node.right))`.
+    * `eval(leaf)` = `leaf.value`.
+  * `schedule` is an implementation-defined full binary tree whose in-order
+    traversal consists of:
+    * `input_slices[:][index]` values, for all `index` in the index space
+      of `input_slices`, in the ascending lexicographic order of `index`.
+    * Interspersed with an implementation-defined amount of `init_values`
+      at implementation-defined positions.
+
+### Inputs
+
+| Name          | Type                                             |
+|---------------|--------------------------------------------------|
+| `inputs`      | variadic number of tensors of any supported type |
+| `init_values` | 0-dimensional tensor of any supported type       |
+| `dimensions`  | 1-dimensional tensor constant of type `si64`     |
+| `body`        | `function`                                       |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `results` | variadic number of tensors of any supported type |
+
+### Constraints
+
+  * (C1) All `inputs` have the same shape.
+  * (C2) element_type(`inputs[k]`) $=$ element_type(`init_values[k]`) $=$
+  element_type(`results[k]`) for all `k` $\in$ [0, N).
+  * (C3) size(`inputs`) $=$ size(`init_values`) $=$ size(`results`) $=$ N where
+  N >= 1.
+  * (C4) 0 $\le$ `dimensions[d]` $\lt$ rank(`inputs[0][d]`) for all dimension
+  `d`.
+  * (C5) All dimensions in `dimensions` are unique.
+  * (C6) The full function type of `body` is
+  `(T0, ..., TN-1, T0, ..., TN-1) -> (T0, ..., TN-1)`, where the first and
+  second set of input types `T0,..., TN-1` corresponds to the element types of
+  `inputs` and `init_values` respectively. The output types `(T0, ..., TN-1)`
+  corresponds to the element types of `results`.
```

Comment:
C6 is no longer needed now that we have C8. Given that, let's overwrite C6 with C8, so that the constraints are ordered in the same way as inputs and outputs are, i.e. C5: dimensions, C6: body, C7: results.